src/backend/updates.rs

```diff
@@ -71,19 +78,93 @@ fn post_upgrade() {
 }
 
 async fn post_upgrade_fixtures() {
+    mutate(|state| state.read_only = true);
+    let total_posts = read(|state| state.posts.len());
+
+    // Init all stable structs becasue we do not need the old contents of the stable memory anymore
+    init_stable_structs();
+
+    loop {
+        let num_posts = read(|state| state.posts.len());
+        let _: Result<(), ()> = canisters::call_canister(id(), "migrate_posts", ((),))
+            .await
+            .map_err(|(code, err)| {
+                mutate(|state| {
+                    state.logger.error(format!(
+                        "couldn't migrate posts: code {:?}, err {}",
+                        code, err
+                    ))
+                })
+            });
+        let new_num_posts = read(|state| state.posts.len());
+        if new_num_posts == 0 {
+            // We're done with migration
+            break;
+        }
+        if num_posts == new_num_posts && num_posts > 0 {
+            // Something went wrong, so we stop and let the state locked.
+            mutate(|state| {
+                state
+                    .logger
+                    .error("couldn't migrate posts: something went wrong")
+            });
+            return;
+        }
+    }
+
+    if total_posts as u64 == Post::count() {
+        mutate(|state| {
+            state.read_only = false;
+            state
+                .logger
+                .debug(format!("Succefully migrated {} posts", total_posts))
+        })
+    }
+}
+
+#[update]
+fn migrate_posts() {
     mutate(|state| {
-        let num_posts = Post::count(state);
-        for id in 0..state.next_post_id {
-            Post::mutate(state, &id, |post| {
-                post.archived = false;
-                Ok(())
-            })
-            .unwrap();
+        let mut r = 0;
```

Comment:
How many instructions does this loop take at most locally? Are we far away from 20B. In terms of performance it might be better to keep this number below 2B (e.g. 1.5B).

Reply:
I didn't measure yet, but serializing 20k posts in one single call that does not do anything is not an issue currently. I hope stable-structs lib does not make it much worse :). But for local tests, I'll add a print line with the number of instructions - thanks for the idea!